ed25519jni/src/main/rust/utils/exception.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use failure::Error;
+use jni::JNIEnv;
+use std::any::Any;
+use std::thread;
+
+type ExceptionResult<T> = thread::Result<Result<T, Error>>;
+
+// Returns value or "throws" exception. `error_val` is returned, because exception will be thrown
+// at the Java side. So this function should be used only for the `panic::catch_unwind` result.
+pub fn unwrap_exc_or<T>(env: &JNIEnv, res: ExceptionResult<T>, error_val: T) -> T {
+    match res {
+        Ok(val) => {
+            match val {
+                Ok(val) => val,
+                Err(jni_error) => {
+                    // Do nothing if there is a pending Java-exception that will be thrown
+                    // automatically by the JVM when the native method returns.
+                    if !env.exception_check().unwrap() {
+                        // Throw a Java exception manually in case of an internal error.
+                        throw(env, &jni_error.to_string())
+                    }
+                    error_val
+                }
+            }
+        }
+        Err(ref e) => {
+            throw(env, &any_to_string(e));
+            error_val
+        }
+    }
```

Comment:
You can use more complicated pattern in the first match to avoid needing the second match:

```suggestion
        Ok(Ok(val)) => val,
        Ok(Err(jni_error)) => {
            // Do nothing if there is a pending Java-exception that will be thrown
            // automatically by the JVM when the native method returns.
            if !env.exception_check().unwrap() {
                // Throw a Java exception manually in case of an internal error.
                throw(env, &jni_error.to_string())
            }
            error_val
        }
        Err(ref e) => {
            throw(env, &any_to_string(e));
            error_val
        }
    }
```